Tuples enhancement

lukedirtwalker opened the pull request:
This is what we discussed this morning.

Note that TupleCondition will go away in a later PR. As if remove currently that would just create more work (i.e. conflicts when rebasing my other stuff).
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746109%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746327%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746331%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746343%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746520%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746717%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746946%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38747138%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38747181%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23issuecomment-137727532%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23issuecomment-137727532%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Thanks%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A48%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/dataformat/TupleSet.cpp%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746327%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20is%20there%20an%20extra%20%60InformationScripting%60%20namespace%20here%3F%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A30%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/TupleSet.cpp%3AL30-66%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/dataformat/TupleSet.cpp%2019%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746331%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20is%20there%20an%20extra%20%60InformationScripting%60%20namespace%20here%3F%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A30%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/TupleSet.cpp%3AL30-66%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/queries/AstQuery.cpp%2031%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746946%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Oh%2C%20that%27s%20an%20issue.%20Please%20think%20about%20what%27s%20the%20best%20thing%20to%20do%20here%2C%20but%20I%20guess%20we%20might%20need%20some%20way%20of%20translating%20these%20connections.%20In%20any%20case%2C%20the%20previous%20solution%20wasn%27t%20entire%20correct%2C%20since%20a%20link%20between%20two%20methods%20%2C%20i.e.%20%5C%22overrides%5C%22%20would%20have%20been%20translated%20to%20a%20link%20between%20two%20classes%2C%20and%20it%20wouldn%27t%20really%20make%20sense%20in%20that%20context.%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A39%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Urgh%20this%20comment%20is%20still%20from%20graph%20times.%20With%20tuples%20the%20connections%20will%20persist%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A42%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL171-185%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/dataformat/TupleSet.cpp%2037%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746343%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20is%20there%20an%20extra%20%60InformationScripting%60%20namespace%20here%3F%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A30%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22sorry%20those%20slipped%20in.%20I%20will%20remove%20them%20everywhere%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A33%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/TupleSet.cpp%3AL30-66%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/dataformat/TupleSet.h%2041%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746717%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Don%27t%20do%20this.%20You%27re%20going%20to%20create%20a%20brand%20new%20QSet%20just%20for%20this%20operation.%20Just%20iterate%20directly%20over%20the%20various%20tuple%20sets%2C%20one%20by%20one.%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A36%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/TupleSet.h%3AL36-107%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/dataformat/Tuple.cpp%2020%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38746109%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20should%20use%20your%20%60find%60%20method.%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A27%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/Tuple.cpp%3AL38-71%22%7D%2C%20%22Pull%2061d71fede3da5e2e7f86926f1683800a762f763f%20InformationScripting/src/visualization/DefaultVisualizer.cpp%2036%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/142%23discussion_r38747138%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%27re%20calling%20find%20twice%20here.%20Don%27t%20do%20that%2C%20just%20remember%20the%20iterator%20and%20use%20it%20later.%22%2C%20%22created_at%22%3A%20%222015-09-04T12%3A41%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/visualization/DefaultVisualizer.cpp%3AL80-103%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/dataformat/Tuple.cpp 20'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38746109'>File: InformationScripting/src/dataformat/Tuple.cpp:L38-71</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> This should use your `find` method.
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/dataformat/TupleSet.cpp 8'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38746327'>File: InformationScripting/src/dataformat/TupleSet.cpp:L30-66</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Why is there an extra `InformationScripting` namespace here?
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/dataformat/TupleSet.cpp 19'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38746331'>File: InformationScripting/src/dataformat/TupleSet.cpp:L30-66</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Why is there an extra `InformationScripting` namespace here?
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/dataformat/TupleSet.cpp 37'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38746343'>File: InformationScripting/src/dataformat/TupleSet.cpp:L30-66</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Why is there an extra `InformationScripting` namespace here?
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> sorry those slipped in. I will remove them everywhere
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/dataformat/TupleSet.h 41'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38746717'>File: InformationScripting/src/dataformat/TupleSet.h:L36-107</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Don't do this. You're going to create a brand new QSet just for this operation. Just iterate directly over the various tuple sets, one by one.
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/queries/AstQuery.cpp 31'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38746946'>File: InformationScripting/src/queries/AstQuery.cpp:L171-185</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Oh, that's an issue. Please think about what's the best thing to do here, but I guess we might need some way of translating these connections. In any case, the previous solution wasn't entire correct, since a link between two methods , i.e. "overrides" would have been translated to a link between two classes, and it wouldn't really make sense in that context.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Urgh this comment is still from graph times. With tuples the connections will persist
- [x] <a href='#crh-comment-Pull 61d71fede3da5e2e7f86926f1683800a762f763f InformationScripting/src/visualization/DefaultVisualizer.cpp 36'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#discussion_r38747138'>File: InformationScripting/src/visualization/DefaultVisualizer.cpp:L80-103</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> You're calling find twice here. Don't do that, just remember the iterator and use it later.
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/142#issuecomment-137727532'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Thanks

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/142?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/142?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/142'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

lukedirtwalker committed:
Only allow unique names in Tuple

This makes the interface nice, and requires one to give more descriptive
names to properties in a tuple.

lukedirtwalker committed:
TupleSet: change the tuples method to take any functor

lukedirtwalker committed:
Store tuples depending on the tag in TupleSet

lukedirtwalker committed:
Add a take method to TupleSet

lukedirtwalker committed:
Add takeAll to TupleSet as well

Remove default argument from templated tuples method, it isn't
applicable and there is a special overload for no arguments.

lukedirtwalker committed:
Remove get methods from Tuple

Instead offer a function to get all values of a certain type.

dimitar-asenov commented on InformationScripting/src/dataformat/Tuple.cpp:
```diff
@@ -38,32 +38,33 @@ QString Tuple::tag() const
-	for (auto p : values_)
-		if (p.first == key) result << p;
-	return result;
+	for (const auto& np : values_)
```
This should use your `find` method.

dimitar-asenov commented on InformationScripting/src/dataformat/TupleSet.cpp:
```diff
@@ -30,21 +30,36 @@ namespace InformationScripting {
+	return tuples_[tag];
+}
+
+QSet<Tuple> InformationScripting::TupleSet::tuples() const
```
Why is there an extra `InformationScripting` namespace here?

dimitar-asenov commented on InformationScripting/src/dataformat/TupleSet.cpp:
```diff
@@ -30,21 +30,36 @@ namespace InformationScripting {
 }
 
-QSet<Tuple> TupleSet::tuples(TupleSet::TupleCondition condition) const
+void InformationScripting::TupleSet::remove(const TupleSet& tuples)
```
Why is there an extra `InformationScripting` namespace here?

dimitar-asenov commented on InformationScripting/src/dataformat/TupleSet.cpp:
```diff
@@ -30,21 +30,36 @@ namespace InformationScripting {
 	return result;
 }
 
+void InformationScripting::TupleSet::unite(const TupleSet& with)
```
Why is there an extra `InformationScripting` namespace here?

lukedirtwalker replied:
sorry those slipped in. I will remove them everywhere

lukedirtwalker committed:
Refactor: Remove unneeded InformationScripting:: specifier

Also change Tuple::contains to use the find method.

dimitar-asenov commented on InformationScripting/src/dataformat/TupleSet.h:
```diff
@@ -36,25 +36,71 @@ class INFORMATIONSCRIPTING_API TupleSet
+	if (!condition) return tuples();
+
+	QSet<Tuple> result;
+	for (auto t : tuples())
```
Don't do this. You're going to create a brand new QSet just for this operation. Just iterate directly over the various tuple sets, one by one.

dimitar-asenov commented on InformationScripting/src/queries/AstQuery.cpp:
```diff
@@ -172,15 +171,14 @@ TupleSet AstQuery::toClassNode(QList<TupleSet> input)
 	};
 
-	auto tuples = ts.tuples(canBeInClass);
+	// TODO currently we remove all the converted nodes, this also means we lose all connections:
```
Oh, that's an issue. Please think about what's the best thing to do here, but I guess we might need some way of translating these connections. In any case, the previous solution wasn't entire correct, since a link between two methods , i.e. "overrides" would have been translated to a link between two classes, and it wouldn't really make sense in that context.

lukedirtwalker replied:
Urgh this comment is still from graph times. With tuples the connections will persist

dimitar-asenov commented on InformationScripting/src/visualization/DefaultVisualizer.cpp:
```diff
@@ -80,27 +80,23 @@ void DefaultVisualizer::visualize(const TupleSet& ts)
-				Model::Node* node = astProperty.second;
-				QString& color = colors[node];
-				if (color.isEmpty())
+				if (t.find("color") != t.end())
```
You're calling find twice here. Don't do that, just remember the iterator and use it later.

lukedirtwalker committed:
Refactor: Small cleanups after review

-Be more efficient when getting tuples for a condition in TupleSet by
not creating QSet copies.
-Only call find once in DefaultVisualizer
-Update a commend in AstQuery

dimitar-asenov commented:
Thanks
